File: 0112-path-sum/0112-path-sum.java
class Solution {
    private boolean result = false;

    public boolean hasPathSum(TreeNode root, int targetSum) {
        if (root == null) {
            return false;
        }
        calc(root, root.val, targetSum);
        return result;
    }

    private void calc(TreeNode node, int total, int targetSum) {
        if (node.left == null && node.right == null) {
            if (total == targetSum) {
                result = true;
                return;
            }
        }
        if (node.left != null) {
            calc(node.left, total + node.left.val, targetSum);
        }
        if (node.right != null) {
            calc(node.right, total + node.right.val, targetSum);
        }
    }
}

File: 0112-path-sum/0112-path-sum.py
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        def calc(node, total):
            if not node.left and not node.right:
                if total == targetSum:
                    nonlocal result
                    result = True
                return
            if node.left:
                calc(node.left, total + node.left.val)
            if node.right:
                calc(node.right, total + node.right.val)


        if root is None: return False

        result = False
        calc(root, root.val)
        return result